gateway/src/apicast/policy/soap/policy.lua

```diff
@@ -0,0 +1,68 @@
+--- SOAP Policy
+-- This policy adds support for a very small subset of SOAP.
+-- This policy basically expects a SOAPAction URI in the SOAPAction header or
+-- the content-type header.
+-- The SOAPAction header is used in v1.1:
+-- https://www.w3.org/TR/2000/NOTE-SOAP-20000508/#_Toc478383528, whereas the
+-- Content-Type header is used in v1.2:
+-- https://www.w3.org/TR/soap12-part2/#ActionFeature
+-- The SOAPAction URI is matched against the mapping rules defined for the
+-- service and the usage resulting from that is authorized and reported
+-- against 3scale's backend.
+
+local sub = string.sub
+local len = string.len
+
+local policy = require('apicast.policy')
+local _M = policy.new('SOAP policy')
+
+local soap_action_header = 'SOAPAction'
+local soap_action_ctype = 'application/soap+xml?action='
+
+local new = _M.new
+
+local function starts_with(str, start)
+  return sub(str, 1, len(start)) == start
+end
+
+local function soap_action_in_header(headers)
+  return headers[soap_action_header]
+end
+
+local function soap_action_in_ctype(headers)
+  local ctype = headers['Content-Type']
+
+  if ctype and starts_with(ctype, soap_action_ctype) then
+    return sub(ctype, len(soap_action_ctype) + 1, -1)
+  else
+    return nil
+  end
+end
+
+local function extract_soap_uri()
+  local headers = ngx.req.get_headers() or {}
+  return soap_action_in_ctype(headers) or soap_action_in_header(headers)
+end
+
+--- Initialize a SOAP policy
+-- @tparam[opt] table config
+function _M.new(config)
+  return new(config)
+end
+
+--- Rewrite phase
+-- Adds to the field 'add_to_usage' of the shared context a table that
+-- contains pairs of (metric, usage).
+-- @tparam table context Shared context between policies
+function _M:rewrite(context)
+  local soap_action_uri = extract_soap_uri()
+
+  if soap_action_uri then
+    local soap_usage = context.service:get_usage(
+      ngx.req.get_method(), soap_action_uri)
+
+    context.add_to_usage = soap_usage
```

Comment:
@mikz I considered several options here but this seemed the cleanest one. We could pass the URI instead and do the mapping rules matching later.

Reply:
I wonder how several policies would interact. Or if this would be twice in the chain.

My idea was to make each policy to increment the values in the context. That way the policy can decide what to do: add, replace, remove, ...

Reply:
Definitely. If in the future we have more policies doing the same thing, we'll need to update `add_to_usage` instead of simply assign it.

Reply:
What I mean is that we can just `context.usage.hits = (context.usage.hits or 0) + 1` in the rewrite phase. Or expose some table merger function to merge those. Or have them in an array and merge them in the proxy:access. It just does not feel right to have `add_to_usage` when the `usage` key is there too.